crates/iceberg/src/arrow.rs

```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+                format!("Unsupported binary operator: {op}"),
+            )),
+        }
+    }
+
+    fn visit_set(&mut self, predicate: &SetExpression<BoundReference>) -> Result<Self::T> {
+        match predicate.op() {
+            PredicateOperator::In => {
+                todo!("In is not supported yet")
+            }
+            PredicateOperator::NotIn => {
+                todo!("NotIn is not supported yet")
+            }
+            op => Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!("Unsupported set operator: {op}"),
+            )),
+        }
+    }
+
+    fn and(&mut self, mut predicates: Vec<Self::T>) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            move |batch| {
+                let left = predicates.get_mut(0).unwrap().evaluate(batch.clone())?;
+                let right = predicates.get_mut(1).unwrap().evaluate(batch)?;
+                and(&left, &right)
+            },
+        )))
+    }
+
+    fn or(&mut self, mut predicates: Vec<Self::T>) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            move |batch| {
+                let left = predicates.get_mut(0).unwrap().evaluate(batch.clone())?;
+                let right = predicates.get_mut(1).unwrap().evaluate(batch)?;
+                or(&left, &right)
+            },
+        )))
+    }
+
+    fn not(&mut self, mut predicate: Self::T) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            move |batch| {
+                let evaluated = predicate.evaluate(batch.clone())?;
+                not(&evaluated)
+            },
+        )))
+    }
+
+    fn bound_reference(&mut self, reference: &BoundReference) -> Result<Self::U> {
+        let column_idx = self.column_map.get(&reference.field().id).ok_or_else(|| {
+            Error::new(
+                ErrorKind::DataInvalid,
+                format!("Field id {} not found in schema", reference.field().id),
+            )
+        })?;
+
+        let root_col_index = self.parquet_schema.get_column_root_idx(*column_idx);
+
+        // Find the column index in projection mask.
+        let column_idx = self
+            .columns
+            .iter()
+            .position(|&x| x == root_col_index)
+            .ok_or_else(|| {
+                Error::new(
+                    ErrorKind::DataInvalid,
+                    format!("Column index {} not found in schema", root_col_index),
+                )
+            })?;
+
+        Ok(column_idx)
+    }
+}
+
+/// A visitor for bound predicates.
+pub trait BoundPredicateVisitor {
+    /// Return type of this visitor on bound predicate.
+    type T;
+
+    /// Return type of this visitor on bound reference.
+    type U;
+
+    /// Visit a bound predicate.
+    fn visit_predicate(&mut self, predicate: &BoundPredicate) -> Result<Self::T> {
```

Comment:
Though it's correct, it's not typical post order visitor pattern, how about moving this travels logic into a standalone   funtion, so that we can follow convention of other visitors?